huntsman/observatory.py

```diff
@@ -512,3 +378,157 @@ def _create_flat_field_observation(self, alt=None, az=None,
+                # Account for the sky changing brightness
+                sky_factor = 2.0 ** (elapsed_time / 180.0)
+                if self.past_midnight:
+                    exptime = exptime / sky_factor
+                else:
+                    exptime = exptime * sky_factor
+                exptime = int(max(min_exptime, exptime))
+                exptimes[cam_name].append(exptime * u.second)
+                self.logger.debug("Elapsed time: {}".format(elapsed_time))
+                self.logger.debug("Suggested exptime for {}: {}".format(cam_name, exptime))
+
+            # Stop flats if any time is greater than max
+            exptime_too_long = [t[-1].value >= max_exptime for t in exptimes.values()]
+            if any(exptime_too_long) and not self.past_midnight:
+                self.logger.debug("Exposure times greater than max, stopping flat fields.")
+                break
+
+            # Stop flats if we have taken too many.
+            if any([len(t) >= max_num_exposures for t in exptimes.values()]):
+                self.logger.debug("Too many flats, quitting.")
+                break
+
+            # Handle saturated cameras
+            if is_saturated and exptimes[cam_name][-1].value < min_saturated_exptime:
+                if self.past_midnight:
+                    self.logger.debug('Saturated with short exposure. Stopping flat fields.')
+                    break
+                else:
+                    self.logger.debug('Saturated with short exposure. Waiting 30 seconds'
+                                      'before next exposure.')
+                    max_num_exposures += 1
+                    time.sleep(30)
+
+        # Return the exposure times
+        return exptimes
+
+    def _take_flat_fields(self, exptimes, observation, fits_headers=None, dark=False):
+        """
+        Slew to flat field, take exposures and wait for them to complete.
+        Returns a list of camera events for each camera.
+
+        args:
+            exptimes: dict of camera_name: list of exposure times.
+            observation: Flat field Observation object.
+        """
+        image_dir = self.config['directories']['images']
+        imtype = 'dark' if dark else 'flat'
+        if fits_headers is None:
+            fits_headers = self.get_standard_headers(observation=observation)
+
+        # Slew to field
+        self.logger.debug(f'Slewing to {imtype}-field coords: {observation.field}.')
+        self.mount.set_target_coordinates(observation.field)
+        self.mount.slew_to_target()
+        self.status()  # Seems to help with reading coords
+
+        # Loop over cameras...
+        camera_events = {}
+        for cam_name, exptime in exptimes.items():
+            cam = self.cameras[cam_name]
+
+            # Define filename for flat fields
+            filename = (f'{image_dir}/flats/{cam.uid}/{observation.seq_time}'
+                        f'/{imtype}_{observation.current_exp_num:02d}.{cam.file_extension}')
```

Comment:
At the moment this is necessary because there is no easy way to get the default filename, which is needed to read in the new flat field images and update the exposure times.